base_local_planner/src/map_grid.cpp

```diff
@@ -218,17 +218,48 @@ namespace base_local_planner{
     std::vector<geometry_msgs::PoseStamped> adjusted_global_plan;
     adjustPlanResolution(global_plan, adjusted_global_plan, costmap.getResolution());
 
-    // skip global path points until we reach the border of the local map
+    auto dist_squared_to_front = [&adjusted_global_plan](const double x, const double y) {
+      const double dx = x - adjusted_global_plan.front().pose.position.x;
+      const double dy = y - adjusted_global_plan.front().pose.position.y;
+      return dx * dx + dy * dy;
+    };
+    bool reached_point_away_from_start = false;
+    bool local_goal_in_obstacle = true;
+
+    // In the following loop, we select a point on the global path as the local goal based on which
+    // we compute the "distance to the goal". The following criteria are chosen:
+    // - local goal needs to be on the map (otherwise, goal does not correspond to a cell)
+    // - local goal needs to be on a cell where the cost is not NO_INFORMATION
+    //   (as NO_INFORMATION is treated as obstacle in updatePathCell())
+    //
+    // Additionally, cells with INSCRIBED_INFLATED_OBSTACLE and LETHAL_OBSTACLE cost are also treated as
+    // an obstacle in updatePathCell()
+    // (This is such that narrow pathways are discarded as potential paths to the goal)
+    // So if the neighboring cells of the local goal cell all have cost of INSCRIBED_INFLATED_OBSTACLE
+    // or LETHAL_OBSTACLE, the goal will be considered unreachable (the local planner will reject all trajectories).
+    // In this case, it is however unreasonable for the local planner to give up already when still far away
+    // from the local goal, as the obstacle on it might disappear or the global plan might get updated.
+    // Hence, once we reached a path point significantly far away from the beginning of the path and we
+    // encounter a cell with INSCRIBED_INFLATED_OBSTACLE / LETHAL_OBSTACLE cost, we can stop searching if
+    // a local goal significantly far away from the beginning of the path was already found that is not in an obstacle.
     for (unsigned int i = 0; i < adjusted_global_plan.size(); ++i) {
       double g_x = adjusted_global_plan[i].pose.position.x;
       double g_y = adjusted_global_plan[i].pose.position.y;
       unsigned int map_x, map_y;
+      reached_point_away_from_start = reached_point_away_from_start || dist_squared_to_front(g_x, g_y) > 1;
       if (costmap.worldToMap(g_x, g_y, map_x, map_y) && costmap.getCost(map_x, map_y) != costmap_2d::NO_INFORMATION) {
+		const bool in_obstacle = costmap.getCost(map_x, map_y) == costmap_2d::INSCRIBED_INFLATED_OBSTACLE ||
+		    costmap.getCost(map_x, map_y) == costmap_2d::LETHAL_OBSTACLE;
```

Comment:
```suggestion
		const bool in_obstacle = costmap.getCost(map_x, map_y) >= costmap_2d::INSCRIBED_INFLATED_OBSTACLE;
```
S1: We know the cost isn't `costmap_2d::NO_INFORMATION` otherwise we wouldn't have entered the `if` statement, so we can simplify this.

S2: Would be nice to store the value of `costmap.getCost(map_x, map_y)` before the `if` statement to avoid calling it multiple times, but that would require changing the structure a bit to handle when `costmap.worldToMap` fails.

Q / F: Can you verify if the indentation for lines 251~252 is correct? When I checked your branch locally the indentation was wrong, but in github it seems ok? :thinking: I think you maybe inserted a `tab` character in line 251 instead of spaces?

Reply:
S1: agreed
S2: yeah I thought that too, but the restructuring needed seems less readable to me
Q/F: we don't have automatic formatting in this repo; it was a tab character